File: golang-devildyw/middle/L77组合.go
package main

func main() {
	combine(5, 4)
}

func combine(n int, k int) [][]int {
	result := make([][]int, 0)
	path := make([]int, 0)

	var dfs func(index int)
	dfs = func(index int) {
		if len(path) == k {
			temp := make([]int, len(path))
			copy(temp, path)
			result = append(result, temp)
		}

		for i := index; i <= n; i++ {
			path = append(path, i)
			dfs(i + 1)
			path = path[:len(path)-1]
		}
	}
	dfs(1)
	return result
}

File: java-devildyw/src/main/java/top/dyw/leetcode/medium/L77组合.java
package top.dyw.leetcode.medium;

import java.util.ArrayList;
import java.util.List;

public class L77组合 {
    public List<List<Integer>> combine(int n, int k) {
        List<List<Integer>> result = new ArrayList();
        List<Integer> path = new ArrayList();

        dfs(1,n, k, path, result);

        return result;

    }

    public void dfs(int index, int n, int k, List<Integer> path, List<List<Integer>> result) {
        // 剩下的长度不够 直接返回
        if (path.size() + (n-index + 1) < k) {
            return;
        }
        if (path.size() == k) {
            result.add(new ArrayList(path));
            return;
        }

        for (int i = index; i<=n; i++) {
            path.add(i);
            dfs(i+1, n, k, path, result);
            path.remove(path.size()-1);
        }
    }
}
